File: dsa-with-subeen-sir/datastructure/stack/stack.py
"""
Stack হলো একটি Linear Data Structure যেখানে
LIFO (Last In First Out) নিয়ম অনুসরণ করা হয়।

LIFO = Last In, First Out

যে element সবার শেষে ঢুকবে (Push),
সেটাই সবার আগে বের হবে (Pop)।
"""

class Stack:
    def __init__(self):
        # create an empty stack
        self.stack = []

    # check stack is empty
    def is_empty(self):
        return len(self.stack) == 0

    def empty(self):
        if self.is_empty():
            print("Stack Underflow!")
            return None

    # Push operation
    def push(self, value):
        # add element at the top (end of list)
        self.stack.append(value)
        print(f"{value} pushed into stack.")

    # Pop operations
    def pop(self):
        # check stack underflow
        self.empty()

        # Remove and return the top element
        return self.stack.pop()

    # Peek Operation
    def peek(self):
        # Check stack is empty
        self.empty()

        # Return top element without removing it
        return self.stack[-1]  # return last element

    def display(self):
        # Check stack is empty
        self.empty()

        print("\nStack Elements (Top to Bottom):")
        for item in reversed(self.stack):
            print(item)

# Driver Code
s = Stack()

s.push(10)
s.push(20)
s.push(30)

s.display()

print("\nTop Element :", s.peek())
print("Popped Element :", s.pop())

s.display()

if s.is_empty():
    print("\nStack is Empty")
else:
    print("\nStack is Not Empty")

File: dsa-with-subeen-sir/datastructure/stack/stack.c
#include <stdio.h>

#define MAX 5

/*
Stack হলো একটি Linear Data Structure যেখানে
LIFO (Last In, First Out) নিয়ম অনুসরণ করা হয়।

LIFO = যে element সবার শেষে ঢুকবে,
সেটাই সবার আগে বের হবে।

Operations:
1. Push    -> নতুন element যোগ করা
2. Pop     -> উপরের element বের করা
3. Peek    -> উপরের element দেখা
4. Display -> সব element দেখানো
*/

typedef struct
{
    int data[MAX];
    int top;
} Stack;

// Initialize Stack
void init(Stack *s)
{
    s->top = -1;
}

// Check if stack is empty
int isEmpty(Stack *s)
{
    return s->top == -1;
}

// Check if stack is full
int isFull(Stack *s)
{
    return s->top == MAX - 1;
}

// Push operation
void push(Stack *s, int value)
{
    if (isFull(s))
    {
        printf("Stack Overflow!\n");
        return;
    }

    s->top++;
    s->data[s->top] = value;

    printf("%d pushed into stack.\n", value);
}

// Pop operation
int pop(Stack *s)
{
    if (isEmpty(s))
    {
        printf("Stack Underflow!\n");
        return -1;
    }

    int value = s->data[s->top];
    s->top--;

    return value;
}

// Peek operation
int peek(Stack *s)
{
    if (isEmpty(s))
    {
        printf("Stack is Empty!\n");
        return -1;
    }

    return s->data[s->top];
}

// Display stack
void display(Stack *s)
{
    if (isEmpty(s))
    {
        printf("Stack is Empty!\n");
        return;
    }

    printf("Stack (Top -> Bottom): ");

    for (int i = s->top; i >= 0; i--)
    {
        printf("%d ", s->data[i]);
    }

    printf("\n");
}

int main()
{
    Stack s;

    init(&s);

    push(&s, 10);
    push(&s, 20);
    push(&s, 30);
    push(&s, 40);
    push(&s, 50);

    display(&s);

    printf("Top Element = %d\n", peek(&s));

    printf("Popped = %d\n", pop(&s));
    printf("Popped = %d\n", pop(&s));

    display(&s);

    printf("Top Element = %d\n", peek(&s));

    return 0;
}